### PR TITLE
Support multiple Credential factories

### PR DESCRIPTION
**Pull request by jkosternl**
Requests are processed in the order of how the credential factories are specified. 
Please verify in the unit tests if I understood every right.

### CODE REVIEW

**Review by tnleeuw — CHANGES_REQUESTED** [review]
Some small changes / improvements which I think can be made.

File: credentialProvider/src/main/java/org/frankframework/credentialprovider/CredentialFactory.java [hunk 1]
```diff
@@ -15,27 +15,29 @@
...
+	private static final String CREDENTIAL_FACTORY_OPTIONAL_PREFIX_KEY = "credentialFactory.optionalPrefix";
+	private static final String DEFAULT_CREDENTIAL_FACTORY = FileSystemCredentialFactory.class.getName();
 
 	public static final String LEGACY_PACKAGE_NAME = "nl.nn.credentialprovider.";
 	public static final String ORG_FRANKFRAMEWORK_PACKAGE_NAME = "org.frankframework.credentialprovider.";
 
 	private static String optionalPrefix;
 
-	private ICredentialFactory delegate;
+	private final LinkedList<ICredentialFactory> delegates = new LinkedList<>();
```
[the remark this answers is not in the document]
Comment (jkosternl) [reply credentialProvider/src/main/java/org/frankframework/credentialprovider/CredentialFactory.java:40]: I guess I was too quick here..an `ArrayList` indeed also gives insertion order guarantee, which is perfect enough. I'll change it.

File: credentialProvider/src/main/java/org/frankframework/credentialprovider/CredentialFactory.java [hunk 2]
```diff
@@ -47,75 +49,92 @@ public class CredentialFactory {
...
+			delegate.initialize();
+			log.info("installed CredentialFactory [" + factoryClassName + "]");
+			delegates.add(delegate);
+		} catch (Exception e) {
+			log.warning("Cannot instantiate CredentialFactory [" + factoryClassName + "] (" + e.getClass().getTypeName() + "): " + e.getMessage());
 		}
-		return false;
 	}
 
 	private static String findAlias(String rawAlias) {
```
[the remark this answers is not in the document]
Comment (jkosternl) [reply credentialProvider/src/main/java/org/frankframework/credentialprovider/CredentialFactory.java:103]: Yes, it was already like that, but I agree that's a better name. I'll change it.

File: credentialProvider/src/main/java/org/frankframework/credentialprovider/CredentialFactory.java [hunk 3]
```diff
@@ -47,75 +49,92 @@ public class CredentialFactory {
...
+		return false;
 	}
 
 	public static ICredentials getCredentials(String rawAlias, Supplier<String> defaultUsernameSupplier, Supplier<String> defaultPasswordSupplier) {
-		ICredentialFactory delegate = getInstance().delegate;
-		if (delegate!=null) {
-			ICredentials result = delegate.getCredentials(findAlias(rawAlias), defaultUsernameSupplier, defaultPasswordSupplier);
-			if (result!=null) {
+		for (ICredentialFactory factory : getInstance().delegates) {
+			ICredentials result = factory.getCredentials(findAlias(rawAlias), defaultUsernameSupplier, defaultPasswordSupplier);
```
[the remark this answers is not in the document]
Comment (jkosternl) [reply credentialProvider/src/main/java/org/frankframework/credentialprovider/CredentialFactory.java:121]: Yes, of course.. I'll change it.

**Review by tnleeuw — APPROVED** [review]

File: credentialProvider/src/test/java/org/frankframework/credentialprovider/CredentialFactoryTest.java [hunk 4]
```diff
@@ -29,4 +41,68 @@ public void testFindAliasWithPrefix() {
...
+		// Act
+		ICredentials c = CredentialFactory.getCredentials("account", null, null);
+
+		// Assert
+		assertEquals("fakeUsername", c.getUsername());
+		assertEquals("fakePassword", c.getPassword()); // comes from property file, ignore mock value
+		assertEquals("[alias1, account, alias2, TheMaster, TheBachelor]", CredentialFactory.getConfiguredAliases().toString());
+		assertTrue(CredentialFactory.hasCredential("TheMaster"));
+		assertTrue(CredentialFactory.hasCredential("fakePrefix:TheBachelor"));
+
```
Comment (nielsm5) [inline credentialProvider/src/test/java/org/frankframework/credentialprovider/CredentialFactoryTest.java:106]: Can you add a test to see what happens when you have 2 CredentialFactories with the same alias?

Comment (jkosternl) [reply credentialProvider/src/test/java/org/frankframework/credentialprovider/CredentialFactoryTest.java]: I think it was already covered by checking `account` which was in both Factories. But I've added some more accounts and data to validate it really works. Also added the check on password-only credential.

**Review by nielsm5 — APPROVED** [review]

**Pull request merged**